sam_view.c

```diff
@@ -817,6 +825,37 @@ static int getLength(char **s)
     return n;
 }
 
+void copy_tag(const char *tag, const bam1_t *rec, kstring_t *linebuf)
+{
+    uint8_t *s = bam_aux_get(rec, tag);
+    if (s) {
+        char aux_type = *s;
+        switch (aux_type) {
+            case 'C':
+            case 'I':
+            case 'S': aux_type = 'I'; break;
+            case 'c':
+            case 's': aux_type = 'i'; break;
+            case 'd': aux_type = 'f'; break;
+        }
+
+        kputc('\t', linebuf);
+        kputsn(tag, 2, linebuf);
+        kputc(':', linebuf);
+        kputc(aux_type=='I'? 'i': aux_type, linebuf);
+        kputc(':', linebuf);
+        switch (aux_type) {
+            case 'H':
+            case 'Z': kputs(bam_aux2Z(s), linebuf); break;
+            case 'i': kputw(bam_aux2i(s), linebuf); break;
+            case 'I': kputuw(bam_aux2i(s), linebuf); break;
+            case 'A': kputc(bam_aux2A(s), linebuf); break;
+            case 'f': kputd(bam_aux2f(s), linebuf); break;
+            default:  kputs("*** Unknown Type ***", linebuf); break;
```

Comment:
This isn't an appropriate way of returning an error.  It ought to be non-void and return -1 or similar.

It also doesn't handle the ```B``` aux type (which in itself is a monstrosity to handle).  If you wish to not handle this particular type then it should be in a comment to state this and also documented.  Note this isn't just a case of not being able to output it if asked for, but any record containing this type produces an error irrespective of whether we wanted to emit this tag.  Even so, it's a big improvement on the previous version which simply core dumped!

Given the complexity of the B tag type I'm wondering if what we really need is to break htslibs ```sam_format1()``` function in two, adding a ```sam_serialise_aux()``` function there, and to then call that function from here to avoid code duplication.

This could make the code more efficient too.  Instead of having a list of tags to copy and for each tag type scanning through the bam aux data to find that tag, we'd stream through the tag list (as done in ```sam_format1```) and either serialise the tag or skip it depending on whether the tag is present in the list of tags to copy (a quick hash lookup).  **I'd be happy with revamping this though to be a subsequent PR**(if ever required0 so for now if you want to get on just make 'B' tag an appropriate warning message about not being supported and raise the issue of B being unsupported as a wishlist item (it's unlikely anyone would actually care for it!).